### PR TITLE
Add SolverName and SolverFamily enums, use them for rule set resolution

### PR DESCRIPTION
**Pull request by gskorokhod**
This PR:

- Adds SolverName and SolverFamily enums 
- Allows to specify the solver families and solver names that a given RuleSet would apply to in the macro
- Adds methods to resolve rule sets based on the target solver / solver family
- Refactors the RuleSet code to move dependency resolution to the RuleSet implementation itself (not a separate module in conjure_oxide) and to encapsulate things better

It is based on #266 

This is part of my work on creating a robust `Stats` struct that would store solver-specific and general stats in the global context.
However, this PR is a small separate unit of work that can be merged "as is"

### CODE REVIEW

**Comment by niklasdewally** [comment]
@gskorokhod What are your lint settings? As far as clippy told me, everything was fine before.
Your settings seem good though, so might be worth adding them to the Cargo.toml?

File: crates/conjure_core/src/solvers.rs [hunk 1]
```diff
@@ -2,32 +2,32 @@ use strum_macros::Display;
 use strum_macros::{EnumIter, EnumString};
 
 /// All supported solvers.
-#[derive(Debug, EnumString, EnumIter, Display)]
-pub enum Solver {
+#[derive(Debug, EnumString, EnumIter, Display, PartialEq, Eq, Hash, Clone, Copy)]
+pub enum SolverName {
     Minion,
     KissSAT,
 }
```
Comment (niklasdewally) [inline crates/conjure_core/src/solvers.rs:9]: I don't think this solver enum is needed or desirable anymore - its just leftover stuff from the older solver interface.

To reduce boilerplate, maybe SolverAdaptor could have a method called get_solver_family()?

Comment (niklasdewally) [reply crates/conjure_core/src/solvers.rs:9]: That way everything about a given solver is defined inside the SolverAdaptor itself: with the current approach, implementers will have to edit this file as well as their adaptor and they may forget, etc.

Comment (niklasdewally) [reply crates/conjure_core/src/solvers.rs:9]: Maybe I am misunderstanding this - is there a purpose of SolverName seperate from both family and adaptor?

Comment (gskorokhod) [reply crates/conjure_core/src/solvers.rs:9]: > Maybe I am misunderstanding this - is there a purpose of SolverName separate from both family and adaptor?

It represents a specific solver that we can target. So, not sure - does SolverAdaptor do that too?

I could try to do that but:

- Solver adaptors are part of conjure_oxide so I might not be able to use them in my context definition
- I still want the family to be able to give a list of its member solvers - this is used for resolving rule sets

**Comment by gskorokhod** [comment]
> @gskorokhod What are your lint settings? Clippy told me everything was fine, yet you've found some more issues it seems? Your settings seem good though, so might be worth adding them to the Cargo.toml?

I haven't changed anything in clippy settings!

My clippy suggested the replacing vecs with slices thing, and the rest of the changes should be just me moving things around, manually re-exporting stuff, or manually fixing typos etc

**Comment by gskorokhod** [comment]
As per discussion with Nik, we don't need SolverName and the architecture needs to be redone somewhat

**Comment by gskorokhod** [comment]
> > As per discussion with Nik, we don't need SolverName and the architecture needs to be redone somewhat
> 
> 
> 
> The dependency management will be tricky - good luck :). As discussed in person, we are at a tricky intersection between every module being its own crate and a singular project - to fix this we need to commit to one way or another.

Looks like the context turned out to be the junction point where we have to make this choice 

Basically, with the old structure, we had a cycle:

- Rules take a Model reference, so need to import model
- Model either comes from core or pulls core as a dependency anyway (for the AST)
- Context contains Rules and RuleSets for stats purposes (we need to see which rules are being applied etc)
- Model contains the context 

No matter how you break these components down, it's a cycle (just with extra steps)

So in my draft PR I solved this by making it one monolithic crate but with submodules to split things logically into parts

**Pull request closed without merging**